File: src/main/java/com/spring/principle/examples/roulette/RouletteGame.java
package com.spring.principle.examples.roulette;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Random;
import java.util.TreeMap;
import java.util.concurrent.ConcurrentHashMap;

import lombok.Getter;
import lombok.extern.slf4j.Slf4j;

@Slf4j
@Getter
public class RouletteGame {

    private final List<Roulette> rouletteList;

    // private final Map<Integer, Roulette> soldOuts = new HashMap<>();
    private final Map<Integer, Roulette> soldOuts = new ConcurrentHashMap<>();

    private final Random random = new Random();

    public RouletteGame(List<Roulette> roulettes) {
        this.rouletteList = roulettes;
    }

    public Roulette play() {
        checkPlayAvailable();
        TreeMap<BigDecimal, Roulette> rouletteProbability = getRouletteProbability();
        Roulette roulette = getResult(rouletteProbability);
        decreaseStock(roulette);
        return roulette;
    }

    private TreeMap<BigDecimal, Roulette> getRouletteProbability() {
        BigDecimal remainProbability = getCurrentProbability();
        BigDecimal cumulativeProbability = BigDecimal.ZERO;
        TreeMap<BigDecimal, Roulette> probabilityMap = new TreeMap<>();
        for (Roulette roulette : this.rouletteList) {
            if (roulette.getStocks() != 0) {
                BigDecimal updatedProbability = roulette.getProbability().add(remainProbability);
                cumulativeProbability = cumulativeProbability.add(updatedProbability);
                probabilityMap.put(cumulativeProbability, roulette);
            }
        }
        return probabilityMap;
    }

    private Roulette getResult(TreeMap<BigDecimal, Roulette> rouletteMap) {
        BigDecimal randomValue = BigDecimal.valueOf(random.nextDouble());
        return rouletteMap.higherEntry(randomValue)
            .getValue();
    }

    private void decreaseStock(Roulette roulette) {
        roulette.decreaseStock();
        if (roulette.getStocks() == 0) {
            soldOuts.put(roulette.getScore(), roulette);
        }
    }

    private BigDecimal getCurrentProbability() {
        synchronized (soldOuts) {
            BigDecimal sumSoldOutProbability = this.soldOuts.values()
                .stream()
                .map(Roulette::getProbability)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
            int availableSize = rouletteList.size() - this.soldOuts.size();
            return sumSoldOutProbability.divide(BigDecimal.valueOf(availableSize), 3, RoundingMode.HALF_UP);
        }
    }

    public Integer getCurrentTotalStock() {
        return this.rouletteList
            .stream()
            .map(Roulette::getStocks)
            .reduce(0, Integer::sum);
    }

    private void checkPlayAvailable() {
        synchronized (soldOuts) {
            boolean equals = Objects.equals(this.rouletteList.size(), this.soldOuts.size());
            if (equals) {
                throw new IllegalStateException();
            }
        }
    }

}

File: src/main/java/com/spring/principle/examples/roulette/RouletteGame.kt
package com.spring.principle.examples.roulette

import java.math.BigDecimal
import java.math.RoundingMode
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.collections.HashMap

class RouletteGame(val sectorList: List<Sector>) {

<<<<<<< Updated upstream
    private val soldOuts = ConcurrentHashMap<Int, Sector>()
=======
<<<<<<< Updated upstream
    private val soldOuts = ConcurrentHashMap<Int, Roulette>()
>>>>>>> Stashed changes
    private val random = Random()
=======
//    private val soldOuts = HashMap<Int, Sector>()
    private var soldOuts = mutableListOf<Sector>()
>>>>>>> Stashed changes

    fun play(): Sector {
        checkPlayAvailable()
        val rouletteProbability = getRouletteProbability()
        val roulette = getResult(rouletteProbability)
        decreaseStock(roulette)
        return roulette
    }

<<<<<<< Updated upstream
    private fun getRouletteProbability(): TreeMap<BigDecimal, Sector> {
=======
<<<<<<< Updated upstream
    private fun getRouletteProbability(): TreeMap<BigDecimal, Roulette> {
>>>>>>> Stashed changes
        val remainProbability = getCurrentProbability()
=======
    private fun getRouletteProbability(): TreeMap<BigDecimal, Sector> {
        val rebalancedProbability = getRebalancedProbability()
>>>>>>> Stashed changes
        var cumulativeProbability = BigDecimal.ZERO
        val probabilityMap = TreeMap<BigDecimal, Sector>()

<<<<<<< Updated upstream
        this.sectorList
=======
<<<<<<< Updated upstream
        this.rouletteList
>>>>>>> Stashed changes
            .filter { r -> r.stocks.get() != 0 }
            .forEach { r ->
                val add = r.probability.add(remainProbability)
=======
        this.sectorList
            .filter { sector -> sector.stocks.get() != 0 }
            .forEach { sector ->
                val add = sector.probability.add(rebalancedProbability)
>>>>>>> Stashed changes
                cumulativeProbability = cumulativeProbability.add(add)
                probabilityMap[cumulativeProbability] = sector
            }
        return probabilityMap
    }

<<<<<<< Updated upstream
    private fun getResult(sectorMap: TreeMap<BigDecimal, Sector>): Sector {
=======
<<<<<<< Updated upstream
    private fun getResult(rouletteMap: TreeMap<BigDecimal, Roulette>): Roulette {
>>>>>>> Stashed changes
        val randomValue = BigDecimal.valueOf(random.nextDouble())
        return sectorMap.higherEntry(randomValue).value
    }

    private fun decreaseStock(sector: Sector) {
        sector.decreaseStock()
        if (sector.stocks.get() == 0) {
            soldOuts[sector.score] = sector
        }
    }

    private fun getCurrentProbability(): BigDecimal? {
        val sumSoldOutProbability = soldOuts.values
            .stream()
<<<<<<< Updated upstream
            .map { obj: Sector -> obj.probability }
=======
            .map { obj: Roulette -> obj.probability }
=======
    private fun getRebalancedProbability(): BigDecimal? {
        val sumSoldOutProbability = soldOuts.stream()
            .map { obj: Sector -> obj.probability }
>>>>>>> Stashed changes
>>>>>>> Stashed changes
            .reduce(BigDecimal.ZERO)
            { obj: BigDecimal, agent: BigDecimal? -> obj.add(agent) }
        val availableSize = sectorList.size - soldOuts.size
        return sumSoldOutProbability.divide(BigDecimal.valueOf(availableSize.toLong()), 3, RoundingMode.HALF_UP)
    }

    private fun decreaseStock(sector: Sector) {
        sector.decreaseStock()
        if (sector.stocks.get() == 0) {
            soldOuts.add(sector)
//            soldOuts[sector.score] = sector
        }
    }

    fun currentTotalStock(): Int {
        return this.sectorList
            .stream()
            .map { obj: Sector -> obj.stocks.get() }
            .reduce(0) { a: Int, b: Int -> Integer.sum(a, b) }
    }

    private fun getResult(sectorMap: TreeMap<BigDecimal, Sector>): Sector {
        val random = Random()
        val randomValue = BigDecimal.valueOf(random.nextDouble())
        return sectorMap.higherEntry(randomValue).value
    }

    private fun checkPlayAvailable() {
        if (this.sectorList.size == this.soldOuts.size) {
            throw IllegalStateException()
        }
    }
}